packages/medusa/src/services/payment-collection.ts

```diff
@@ -160,4 +253,479 @@ export default class PaymentCollectionService extends TransactionBaseService {
       return paymentCollection
     })
   }
+
+  private isValidTotalAmount(
+    total: number,
+    sessionsInput: PaymentCollectionSessionInput[]
+  ): boolean {
+    const sum = sessionsInput.reduce((cur, sess) => cur + sess.amount, 0)
+    return total === sum
+  }
+
+  async setPaymentSessions(
+    paymentCollectionId: string,
+    sessions: PaymentCollectionSessionInput[] | PaymentCollectionSessionInput
+  ): Promise<PaymentCollection> {
+    let sessionsInput = Array.isArray(sessions) ? sessions : [sessions]
+
+    return await this.atomicPhase_(async (manager: EntityManager) => {
+      const paymentCollectionRepository = manager.getCustomRepository(
+        this.paymentCollectionRepository_
+      )
+
+      const payCol = await this.retrieve(paymentCollectionId, {
+        relations: ["region", "region.payment_providers", "payment_sessions"],
+      })
+
+      if (payCol.status !== PaymentCollectionStatus.NOT_PAID) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_ALLOWED,
+          `Cannot set payment sessions for a payment collection with status ${payCol.status}`
+        )
+      }
+
+      sessionsInput = sessionsInput.filter((session) => {
+        return !!payCol.region.payment_providers.find(({ id }) => {
+          return id === session.provider_id
+        })
+      })
+
+      if (!this.isValidTotalAmount(payCol.amount, sessionsInput)) {
+        throw new MedusaError(
+          MedusaError.Types.UNEXPECTED_STATE,
+          `The total amount must be equal to ${payCol.amount}`
+        )
+      }
+
+      let customer: Customer | undefined = undefined
+
+      const selectedSessionIds: string[] = []
+      const paymentSessions: PaymentSession[] = []
+
+      for (const session of sessionsInput) {
+        if (!customer) {
+          customer = await this.customerService_.retrieve(session.customer_id, {
```

Comment:
todo: Should use `withTransaction`
```suggestion
          customer = await this.customerService_.withTransaction(manager).retrieve(session.customer_id, {
```